cove_bods/process.py

```diff
@@ -0,0 +1,395 @@
+from libcovebods.schema import SchemaBODS
+from libcovebods.config import LibCoveBODSConfig
+from libcovebods.jsonschemavalidate import JSONSchemaValidator
+from libcovebods.additionalfields import AdditionalFields
+import libcovebods.run_tasks
+import libcovebods.data_reader
+from typing import List
+
+import json
+import os.path
+
+import flattentool
+from sentry_sdk import capture_exception
+
+from libcoveweb2.models import SuppliedDataFile, SuppliedData
+from libcoveweb2.process.base import ProcessDataTask
+from libcoveweb2.process.common_tasks.task_with_state import TaskWithState
+
+# from libcove.lib.converters import convert_json, convert_spreadsheet
+from libcoveweb2.utils import get_file_type_for_flatten_tool
+from libcoveweb2.utils import group_data_list_by
+
+
+class Sample(ProcessDataTask):
+    def is_processing_applicable(self) -> bool:
+        return True
+
+    def process(self, process_data: dict) -> dict:
+        process_data["sample_mode"] = self.supplied_data.meta.get("sample_mode")
+        return process_data
+
+    def get_context(self):
+        return {"sample_mode": self.supplied_data.meta.get("sample_mode")}
+
+
+class SetOrTestSuppliedDataFormat(ProcessDataTask):
+
+    map_file_type_to_format = {
+        'json': 'json',
+        'xlsx': 'spreadsheet',
+        'ods': 'spreadsheet'
+    }
+
+    def is_processing_applicable(self) -> bool:
+        return True
+
+    def is_processing_needed(self) -> bool:
+        return self.supplied_data.format == "unknown"
+
+    def process(self, process_data: dict) -> dict:
+        if self.supplied_data.format == "unknown":
+            # Look up what data format is, and set it
+            supplied_data_files = SuppliedDataFile.objects.filter(
+                supplied_data=self.supplied_data
+            )
+            all_file_types = [get_file_type_for_flatten_tool(i) for i in supplied_data_files]
+            file_types_reduced = list(set([i for i in all_file_types if i]))
+            if len(file_types_reduced) == 1:
+                self.supplied_data.format = self.map_file_type_to_format[file_types_reduced[0]]
+                self.supplied_data.save()
+
+            elif len(file_types_reduced) == 0:
+                raise Exception("GOT ZERO")
+                # TODO
+
+            elif len(file_types_reduced) > 1:
+                raise Exception("GOT MORE THAN ONE")
```

Comment:
`NotImplementedError` would be a good choice here.

Reply:
Done